scio-jdbc/src/main/java/com/spotify/scio/jdbc/JdbcDoFn.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2023 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.jdbc;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import javax.sql.DataSource;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+
+/**
+ * A {@link DoFn} that performs synchronous lookup using JDBC connection.
+ *
+ * @param <A> input element type.
+ * @param <B> JDBC lookup value type.
+ */
+public abstract class JdbcDoFn<A, B> extends DoFn<A, B> {
+
+  private final SerializableFunction<Void, DataSource> dataSourceProviderFn;
```

Comment:
I think `Supplier` is more adapted here

Reply:
Neither of available `Supplier` interfaces are serializable

Reply:
ha ok. Maybe we can simply add a 'nicer' API with the Supplier as constructor parameter, and convert it to the serializable function.